File: apps/ios/Sources/Services/OfflineQueue.swift
import Foundation

// MARK: - QueuedOperationType

/// Types of write operations that can be queued for offline replay.
enum QueuedOperationType: String, Codable, Sendable {
    case noteCreate = "note:create"
    case noteUpdate = "note:update"
    case messageSend = "message:send"
    case shiftToggle = "shift:toggle"
    case reportCreate = "report:create"
    case reportMessage = "report:message"
    case banAdd = "ban:add"
    case banRemove = "ban:remove"
    case genericWrite = "generic:write"
}

// MARK: - QueuedOperation

/// A write operation persisted for later replay when the network is available.
struct QueuedOperation: Codable, Identifiable, Sendable {
    let id: String
    let type: QueuedOperationType
    /// API path (e.g., "/api/notes")
    let path: String
    /// HTTP method (POST, PUT, PATCH, DELETE)
    let method: String
    /// JSON-serialized request body, or nil for bodyless requests
    let body: String?
    /// ISO 8601 timestamp when the operation was queued
    let queuedAt: String
    /// Number of replay attempts so far
    var attempts: Int
    /// Last error message from a failed replay attempt
    var lastError: String?
}

// MARK: - OfflineQueue

/// Persists write operations when the network is unavailable and replays them
/// in FIFO order on reconnect with exponential backoff.
///
/// Operations are encrypted at rest using CryptoService and stored to a file
/// in the app's Application Support directory. The queue observes NWPathMonitor
/// for connectivity changes and triggers automatic replay on reconnect.
///
/// Usage:
/// - Wire into APIService: when a write request fails with a network error, call `enqueue(...)`
/// - The queue replays automatically when connectivity is restored
/// - UI observes `pendingCount` to show a banner
@Observable
final class OfflineQueue: @unchecked Sendable {

    // MARK: - Constants

    private static let maxAttempts = 10
    private static let baseRetryDelay: TimeInterval = 1.0
    private static let maxRetryDelay: TimeInterval = 60.0
    private static let queueFileName = "offline-queue.json"

    // MARK: - Observable State

    /// Number of operations waiting to be replayed.
    private(set) var pendingCount: Int = 0

    /// Whether the queue is currently replaying operations.
    private(set) var isReplaying: Bool = false

    // MARK: - Private State

    private var queue: [QueuedOperation] = []
    private let apiService: APIService
    private var networkMonitor: NetworkMonitorService?
    private var connectivityTask: Task<Void, Never>?

    // MARK: - Init

    init(apiService: APIService) {
        self.apiService = apiService
        loadFromDisk()
        pendingCount = queue.count
    }

    // MARK: - Network Monitoring

    /// Start observing network connectivity. Call once after services are ready.
    /// When connectivity is restored and the queue is non-empty, replay is triggered.
    func startMonitoring() {
        let monitor = NetworkMonitorService()
        self.networkMonitor = monitor
        monitor.start()

        connectivityTask?.cancel()
        connectivityTask = Task { [weak self] in
            for await isConnected in monitor.changes {
                guard !Task.isCancelled else { break }
                if isConnected {
                    await self?.replay()
                }
            }
        }
    }

    /// Stop monitoring connectivity.
    func stopMonitoring() {
        connectivityTask?.cancel()
        connectivityTask = nil
        networkMonitor?.stop()
        networkMonitor = nil
    }

    // MARK: - Enqueue

    /// Add a write operation to the queue for later replay.
    ///
    /// - Parameters:
    ///   - path: API path (e.g., "/api/notes")
    ///   - method: HTTP method (POST, PUT, PATCH, DELETE)
    ///   - body: JSON-serialized request body, or nil
    /// - Returns: The queued operation ID
    @discardableResult
    func enqueue(path: String, method: String, body: String?) -> String {
        let op = QueuedOperation(
            id: UUID().uuidString,
            type: Self.classifyOperation(path: path, method: method),
            path: path,
            method: method,
            body: body,
            queuedAt: ISO8601DateFormatter().string(from: Date()),
            attempts: 0,
            lastError: nil
        )
        queue.append(op)
        pendingCount = queue.count
        saveToDisk()
        return op.id
    }

    // MARK: - Replay

    /// Attempt to replay all queued operations in FIFO order.
    ///
    /// Operations are removed on success (HTTP 2xx or 409 conflict).
    /// Permanent client errors (4xx except 401/429) increment the attempt counter;
    /// the operation is removed after maxAttempts.
    /// Server errors (5xx) and network errors cause retry with exponential backoff.
    @MainActor
    func replay() async {
        guard !isReplaying, !queue.isEmpty else { return }

        isReplaying = true
        var toRemove: Set<String> = []

        for i in queue.indices {
            var op = queue[i]

            do {
                let (statusCode, _) = try await executeRequest(op)

                if (200...299).contains(statusCode) || statusCode == 409 {
                    // Success or idempotent conflict — remove
                    toRemove.insert(op.id)
                } else if statusCode >= 400 && statusCode < 500 && statusCode != 401 && statusCode != 429 {
                    // Permanent client error — increment attempts
                    op.attempts += 1
                    op.lastError = "HTTP \(statusCode)"
                    if op.attempts >= Self.maxAttempts {
                        toRemove.insert(op.id)
                    }
                    queue[i] = op
                } else {
                    // Server error / auth / rate limit — retry with backoff
                    op.attempts += 1
                    op.lastError = "HTTP \(statusCode)"
                    queue[i] = op
                    let delay = min(
                        Self.baseRetryDelay * pow(2.0, Double(op.attempts)),
                        Self.maxRetryDelay
                    )
                    try? await Task.sleep(for: .seconds(delay))
                }
            } catch {
                // Network error — stop processing, retry when online
                op.attempts += 1
                op.lastError = error.localizedDescription
                queue[i] = op
                break
            }
        }

        queue.removeAll { toRemove.contains($0.id) }
        pendingCount = queue.count
        saveToDisk()
        isReplaying = false
    }

    // MARK: - Remove / Clear

    /// Remove a single operation from the queue.
    func remove(id: String) {
        queue.removeAll { $0.id == id }
        pendingCount = queue.count
        saveToDisk()
    }

    /// Clear all queued operations.
    func clear() {
        queue.removeAll()
        pendingCount = 0
        saveToDisk()
    }

    /// Current queue snapshot for UI display.
    func getQueue() -> [QueuedOperation] {
        return queue
    }

    // MARK: - Request Execution

    /// Execute a queued operation against the API. Returns (statusCode, responseBody).
    ///
    /// Delegates to `APIService.rawRequest` so the replayed request carries the same
    /// Ed25519 auth token and cert-pinning URLSession as a live request. Using
    /// `URLSession.shared` here would bypass both, causing 401s on every replay.
    private func executeRequest(_ op: QueuedOperation) async throws -> (Int, String) {
        return try await apiService.rawRequest(method: op.method, path: op.path, body: op.body)
    }

    // MARK: - Classification

    /// Classify an API path + method into a queue operation type.
    static func classifyOperation(path: String, method: String) -> QueuedOperationType {
        if path.contains("/notes") && method.uppercased() == "POST" { return .noteCreate }
        if path.contains("/notes/") && method.uppercased() == "PATCH" { return .noteUpdate }
        if path.contains("/messages") && method.uppercased() == "POST" { return .messageSend }
        if path.contains("/shifts/my-status") || path.contains("/shifts/clock") { return .shiftToggle }
        if path.contains("/reports") && method.uppercased() == "POST" && !path.contains("/messages") { return .reportCreate }
        if path.contains("/reports/") && path.contains("/messages") && method.uppercased() == "POST" { return .reportMessage }
        if path.contains("/bans") && method.uppercased() == "POST" { return .banAdd }
        if path.contains("/bans/") && method.uppercased() == "DELETE" { return .banRemove }
        return .genericWrite
    }

    /// Whether an HTTP method is a write operation eligible for queueing.
    static func isQueueableMethod(_ method: String) -> Bool {
        let upper = method.uppercased()
        return upper == "POST" || upper == "PUT" || upper == "PATCH" || upper == "DELETE"
    }

    // MARK: - Persistence

    private var queueFileURL: URL {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let dir = appSupport.appendingPathComponent("llamenos", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent(Self.queueFileName)
    }

    private func saveToDisk() {
        do {
            let data = try JSONEncoder().encode(queue)
            try data.write(to: queueFileURL, options: [.atomic, .completeFileProtection])
        } catch {
            // Persistence failure — operations may be lost on app termination
        }
    }

    private func loadFromDisk() {
        do {
            let data = try Data(contentsOf: queueFileURL)
            queue = try JSONDecoder().decode([QueuedOperation].self, from: data)
        } catch {
            queue = []
        }
    }
}

// MARK: - NetworkMonitorService (NWPathMonitor wrapper)

import Network

/// Lightweight NWPathMonitor wrapper for observing connectivity changes.
/// Used internally by OfflineQueue; not a general-purpose network monitor.
private final class NetworkMonitorService: @unchecked Sendable {
    private let monitor = NWPathMonitor()
    private let monitorQueue = DispatchQueue(label: "org.llamenos.network-monitor")

    /// Current connectivity state (true = reachable).
    private(set) var isConnected: Bool = true

    /// AsyncStream that yields `true` when the path becomes reachable and
    /// `false` when it becomes unreachable. Consumers drive replay on reconnect
    /// without any polling.
    private let _continuation: AsyncStream<Bool>.Continuation
    let changes: AsyncStream<Bool>

    init() {
        (changes, _continuation) = AsyncStream<Bool>.makeStream()
    }

    func start() {
        monitor.pathUpdateHandler = { [weak self] path in
            guard let self else { return }
            let connected = path.status == .satisfied
            self.isConnected = connected
            self._continuation.yield(connected)
        }
        monitor.start(queue: monitorQueue)
    }

    func stop() {
        _continuation.finish()
        monitor.cancel()
    }
}

File: apps/android/app/src/main/java/org/llamenos/hotline/service/OfflineQueue.kt
package org.llamenos.hotline.service

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.llamenos.hotline.api.ApiException
import org.llamenos.hotline.api.ApiService
import org.llamenos.hotline.api.NetworkMonitor
import org.llamenos.hotline.crypto.KeyValueStore
import java.io.IOException
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.min
import kotlin.math.pow

/**
 * Types of write operations that can be queued for offline replay.
 */
@Serializable
enum class QueuedOperationType {
    NOTE_CREATE,
    NOTE_UPDATE,
    MESSAGE_SEND,
    SHIFT_TOGGLE,
    REPORT_CREATE,
    REPORT_MESSAGE,
    BAN_ADD,
    BAN_REMOVE,
    GENERIC_WRITE,
}

/**
 * A write operation persisted for later replay when the network is available.
 */
@Serializable
data class QueuedOperation(
    val id: String,
    val type: QueuedOperationType,
    /** API path (e.g., "/api/notes") */
    val path: String,
    /** HTTP method (POST, PUT, PATCH, DELETE) */
    val method: String,
    /** JSON-serialized request body, or null for bodyless requests */
    val body: String?,
    /** ISO 8601 timestamp when the operation was queued */
    val queuedAt: String,
    /** Number of replay attempts so far */
    var attempts: Int = 0,
    /** Last error message from a failed replay attempt */
    var lastError: String? = null,
)

/**
 * Persists write operations when the network is unavailable and replays them
 * in FIFO order on reconnect with exponential backoff.
 *
 * Operations are stored in [KeyValueStore] (EncryptedSharedPreferences) so they
 * are encrypted at rest. The queue observes [NetworkMonitor] for connectivity
 * changes and triggers automatic replay on reconnect.
 *
 * Injected as a singleton via Hilt. UI observes [pendingCount] to show the
 * offline banner with pending operation count.
 */
@Singleton
class OfflineQueue @Inject constructor(
    private val keystoreService: KeyValueStore,
    private val networkMonitor: NetworkMonitor,
) {

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private var queue: MutableList<QueuedOperation> = mutableListOf()

    private val _pendingCount = MutableStateFlow(0)

    /** Number of operations waiting to be replayed. Observable by UI. */
    val pendingCount: StateFlow<Int> = _pendingCount.asStateFlow()

    private val _isReplaying = MutableStateFlow(false)

    /** Whether the queue is currently replaying operations. */
    val isReplaying: StateFlow<Boolean> = _isReplaying.asStateFlow()

    /**
     * Late-initialized reference to ApiService. Set by AppModule or MainScreen
     * after Hilt wires everything. Avoids circular dependency (ApiService -> OfflineQueue -> ApiService).
     */
    var apiService: ApiService? = null

    init {
        loadFromDisk()
        _pendingCount.value = queue.size
    }

    /**
     * Start observing network connectivity. When connectivity is restored
     * and the queue is non-empty, replay is triggered automatically.
     */
    fun startMonitoring() {
        scope.launch {
            var wasOnline = networkMonitor.isOnline.value
            networkMonitor.isOnline.collect { isOnline ->
                if (isOnline && !wasOnline && queue.isNotEmpty()) {
                    replay()
                }
                wasOnline = isOnline
            }
        }
    }

    /**
     * Add a write operation to the queue for later replay.
     *
     * @param path API path (e.g., "/api/notes")
     * @param method HTTP method (POST, PUT, PATCH, DELETE)
     * @param body JSON-serialized request body, or null
     * @return The queued operation ID
     */
    fun enqueue(path: String, method: String, body: String?): String {
        val op = QueuedOperation(
            id = UUID.randomUUID().toString(),
            type = classifyOperation(path, method),
            path = path,
            method = method.uppercase(),
            body = body,
            queuedAt = java.time.Instant.now().toString(),
        )
        synchronized(queue) {
            queue.add(op)
            _pendingCount.value = queue.size
        }
        saveToDisk()
        return op.id
    }

    /**
     * Attempt to replay all queued operations in FIFO order.
     *
     * Operations are removed on success (HTTP 2xx or 409 conflict).
     * Permanent client errors (4xx except 401/429) increment the attempt counter;
     * the operation is removed after [MAX_ATTEMPTS].
     * Server errors (5xx) and network errors cause retry with exponential backoff.
     */
    suspend fun replay() {
        val api = apiService ?: return
        if (_isReplaying.value || queue.isEmpty()) return

        _isReplaying.value = true
        val toRemove = mutableSetOf<String>()

        val snapshot = synchronized(queue) { queue.toList() }

        for ((index, op) in snapshot.withIndex()) {
            if (!networkMonitor.isOnline.value) break

            try {
                api.requestRawNoContent(op.method, op.path, op.body)
                toRemove.add(op.id)
            } catch (e: ApiException) {
                val code = e.code
                if (code == 409) {
                    // Idempotent conflict — already applied
                    toRemove.add(op.id)
                } else if (code in 400..499 && code != 401 && code != 429) {
                    // Permanent client error
                    synchronized(queue) {
                        queue.getOrNull(index)?.let { mutableOp ->
                            mutableOp.attempts++
                            mutableOp.lastError = "HTTP $code"
                            if (mutableOp.attempts >= MAX_ATTEMPTS) {
                                toRemove.add(mutableOp.id)
                            }
                        }
                    }
                } else {
                    // Server error / auth / rate limit — retry with backoff
                    synchronized(queue) {
                        queue.getOrNull(index)?.let { mutableOp ->
                            mutableOp.attempts++
                            mutableOp.lastError = "HTTP $code"
                        }
                    }
                    val delayMs = min(
                        BASE_RETRY_DELAY_MS * 2.0.pow(op.attempts.toDouble()),
                        MAX_RETRY_DELAY_MS.toDouble(),
                    ).toLong()
                    delay(delayMs)
                }
            } catch (_: IOException) {
                // Network error — stop processing
                synchronized(queue) {
                    queue.getOrNull(index)?.let { mutableOp ->
                        mutableOp.attempts++
                        mutableOp.lastError = "Network error"
                    }
                }
                break
            } catch (_: Exception) {
                // Unexpected error — skip and continue
                synchronized(queue) {
                    queue.getOrNull(index)?.let { mutableOp ->
                        mutableOp.attempts++
                        mutableOp.lastError = "Unexpected error"
                    }
                }
            }
        }

        synchronized(queue) {
            queue.removeAll { it.id in toRemove }
            _pendingCount.value = queue.size
        }
        saveToDisk()
        _isReplaying.value = false
    }

    /**
     * Remove a single operation from the queue.
     */
    fun remove(id: String) {
        synchronized(queue) {
            queue.removeAll { it.id == id }
            _pendingCount.value = queue.size
        }
        saveToDisk()
    }

    /**
     * Clear all queued operations.
     */
    fun clear() {
        synchronized(queue) {
            queue.clear()
            _pendingCount.value = 0
        }
        saveToDisk()
    }

    /**
     * Current queue snapshot for UI display.
     */
    fun getQueue(): List<QueuedOperation> = synchronized(queue) { queue.toList() }

    // MARK: - Classification

    private fun classifyOperation(path: String, method: String): QueuedOperationType {
        val m = method.uppercase()
        return when {
            path.contains("/notes") && m == "POST" -> QueuedOperationType.NOTE_CREATE
            path.contains("/notes/") && m == "PATCH" -> QueuedOperationType.NOTE_UPDATE
            path.contains("/messages") && m == "POST" -> QueuedOperationType.MESSAGE_SEND
            path.contains("/shifts/my-status") || path.contains("/shifts/clock") -> QueuedOperationType.SHIFT_TOGGLE
            path.contains("/reports") && m == "POST" && !path.contains("/messages") -> QueuedOperationType.REPORT_CREATE
            path.contains("/reports/") && path.contains("/messages") && m == "POST" -> QueuedOperationType.REPORT_MESSAGE
            path.contains("/bans") && m == "POST" -> QueuedOperationType.BAN_ADD
            path.contains("/bans/") && m == "DELETE" -> QueuedOperationType.BAN_REMOVE
            else -> QueuedOperationType.GENERIC_WRITE
        }
    }

    // MARK: - Persistence

    private fun saveToDisk() {
        try {
            val data = synchronized(queue) { json.encodeToString(queue.toList()) }
            keystoreService.store(STORAGE_KEY, data)
        } catch (_: Exception) {
            // Persistence failure — operations may be lost on app termination
        }
    }

    private fun loadFromDisk() {
        try {
            val data = keystoreService.retrieve(STORAGE_KEY) ?: return
            queue = json.decodeFromString<List<QueuedOperation>>(data).toMutableList()
        } catch (_: Exception) {
            queue = mutableListOf()
        }
    }

    companion object {
        private const val STORAGE_KEY = "offline-queue"
        private const val MAX_ATTEMPTS = 10
        private const val BASE_RETRY_DELAY_MS = 1000L
        private const val MAX_RETRY_DELAY_MS = 60_000L

        /** Write methods eligible for offline queueing. */
        private val QUEUEABLE_METHODS = setOf("POST", "PUT", "PATCH", "DELETE")

        /** Whether an HTTP method is a write operation eligible for queueing. */
        fun isQueueableMethod(method: String): Boolean =
            method.uppercase() in QUEUEABLE_METHODS
    }
}

